### PR TITLE
feat(autoware_utils_geometry): improve boost_polygon_utils to guard against degenerate inputs

### PR DESCRIPTION
**Pull request by mkquda**
## Description

This PR adds safer polygon conversion/expansion APIs for degenerate perception footprints, while keeping the existing `to_polygon2d()` and `expand_polygon()` behavior unchanged. It also adds buffered `to_polygon2d` overloads for convenience.

### Background
Perception objects can occasionally carry invalid or degenerate `POLYGON` footprints (single point, line segment, duplicate/collinear vertices). Passing these directly into Boost.Geometry operations can produce invalid polygons, unstable expansion, or runtime failures. An example is shown below:

Screenshot from 2026-08-04 10-14-45

This PR introduces `*_safe` variants that callers can opt into when robustness against degenerate inputs is required.

### Changes
- Add `to_polygon2d_safe()`:
  - 1-point footprint → adds two offset vertices (0.05 m) to form a small triangle
  - 2-point footprint → adds a perpendicular offset vertex at the segment midpoint to form a triangle
  - Includes overloads for buffered polygons and perception object types
- Add `expand_polygon_safe()`:
  - Early return for inputs with fewer than 3 vertices or negligible buffer (`offset < 1e-2`)
  - Pre-filters invalid vertices before expansion (duplicate consecutive points, collinear points)
  - Returns the original polygon unchanged if filtering leaves fewer than 3 vertices
  - Guards the expansion loop against remaining degeneracy (zero-length edges, zero offset vector)
- Add unit tests for the safe APIs (degenerate footprints, empty/short/collinear polygons)
- Add `to_polygon2d(..., buffer)` overloads to reduce boilerplate at call sites

## How was this PR tested?
- `colcon build --packages-select autoware_utils_geometry --symlink-install --cmake-args -DCMAKE_BUILD_TYPE=Release`
- `colcon test --packages-select autoware_utils_geometry`

## Notes for reviewers

None.

## Effects on system behavior

None.

### CODE REVIEW

File: autoware_utils_geometry/src/geometry/boost_polygon_utils.cpp [hunk 1]
```diff
@@ -171,6 +171,25 @@ Polygon2d to_polygon2d(
...
+  if (outer.size() == 1) {
+    const auto p = outer.front();
+    append_point_to_polygon(polygon, Point2d(p.x() + dummy_offset, p.y()));
+    append_point_to_polygon(polygon, Point2d(p.x(), p.y() - dummy_offset));
+  } else if (outer.size() == 2) {
+    const auto & p0 = outer.at(0);
+    const auto & p1 = outer.at(1);
+    const double dx = p1.x() - p0.x();
+    const double dy = p1.y() - p0.y();
+    const double len = std::hypot(dx, dy);
```
Comment (yuki-takagi-66) [inline autoware_utils_geometry/src/geometry/boost_polygon_utils.cpp]: Since this function may be called frequently, it would be preferable to avoid heavy functions like hypot. That said, if there are no better alternatives, I think this is acceptable.

File: autoware_utils_geometry/src/geometry/boost_polygon_utils.cpp [hunk 2]
```diff
@@ -171,6 +171,25 @@ Polygon2d to_polygon2d(
...
+    const double mx = 0.5 * (p0.x() + p1.x());
+    const double my = 0.5 * (p0.y() + p1.y());
+    append_point_to_polygon(
+      polygon, Point2d(mx + dy / len * dummy_offset, my - dx / len * dummy_offset));
```
Comment (yuki-takagi-66) [inline autoware_utils_geometry/src/geometry/boost_polygon_utils.cpp]: Even if we have to leave hypot(), zero division handling may be required

File: autoware_utils_geometry/src/geometry/boost_polygon_utils.cpp (hunk 1 above)
Comment (mkquda) [reply autoware_utils_geometry/src/geometry/boost_polygon_utils.cpp]: resolved in c45650c

File: autoware_utils_geometry/src/geometry/boost_polygon_utils.cpp (hunk 2 above)
Comment (mkquda) [reply autoware_utils_geometry/src/geometry/boost_polygon_utils.cpp]: resolved in c45650c

**Review by yuki-takagi-66 — APPROVED** [review]
lgtm

**Review by isamu-takagi — COMMENTED** [review]
This PR changes the behavior of `to_polygon2d`; is that acceptable? For instance, error-handling logic that checks the number of vertices after calling `to_polygon2d` will no longer work.

If such an impact is anticipated, please create the conventional function and the function with the correction feature separately.

**Comment by mkquda** [comment]
@isamu-takagi 
> ` [540e7aa](/autowarefoundation/autoware_utils/pull/117/commits/540e7aaf952d7f1a5fb9dd87e6fd7d6d3f8d6f7c)`

resolved in 540e7aa

**Review by isamu-takagi — APPROVED** [review]
